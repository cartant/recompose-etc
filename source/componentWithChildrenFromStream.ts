import * as React from "react";
import {
  componentFromStreamWithConfig,
  mapper,
  Subscribable
} from "recompose";
import { Observable } from "rxjs/Observable";
import { from } from "rxjs/observable/from";
import { map } from "rxjs/operators/map";
import { withLatestFrom } from "rxjs/operators/withLatestFrom";
import { rxjsObservableConfig } from "./rxjsObservableConfig";

export function componentWithChildrenFromStream<TProps, TChildProps>(
  propsToReactNode: mapper<Subscribable<TProps>, Observable<TChildProps>>
): React.ComponentType<TProps & { children?: (props: TChildProps) => React.ReactNode }> & { ChildProps: TChildProps } {
  const componentFromStream = componentFromStreamWithConfig(rxjsObservableConfig);
  return componentFromStream<TProps & { children?: (props: TChildProps) => React.ReactNode }>(props$ => {
    return from(propsToReactNode(props$)).pipe(
      withLatestFrom(props$),
      map(([childProps, props]) => props.children!(childProps))
    );
  }) as any;
}
