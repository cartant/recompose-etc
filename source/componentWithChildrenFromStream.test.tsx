import { shallow } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { componentWithChildrenFromStream } from "./componentWithChildrenFromStream";
import { counterPropsFromStream } from "./testing/counter";

const Counter = componentWithChildrenFromStream(counterPropsFromStream);
const element = () => (
  <Counter>
    {({ count, increment, decrement }: typeof Counter.ChildProps) => (<div>
      Count: {count}
      <button id="increment" onClick={increment}>+</button>
      <button id="decrement" onClick={decrement}>-</button>
    </div>)}
  </Counter>
);

it("should render without crashing", () => {
  shallow(element());
});

it("should increment the counter", () => {
  const wrapper = shallow(element());
  expect(wrapper.html()).toEqual(expect.stringMatching(/Count: 0/));
  wrapper.find("#increment").simulate("click");
  expect(wrapper.html()).toEqual(expect.stringMatching(/Count: 1/));
});

it("should decrement the counter", () => {
  const wrapper = shallow(element());
  expect(wrapper.html()).toEqual(expect.stringMatching(/Count: 0/));
  wrapper.find("#decrement").simulate("click");
  expect(wrapper.html()).toEqual(expect.stringMatching(/Count: -1/));
});

it("should render correctly", () => {
  const rendering = renderer.create(element()).toJSON();
  expect(rendering).toMatchSnapshot();
});
