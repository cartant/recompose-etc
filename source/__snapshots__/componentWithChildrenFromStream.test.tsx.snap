// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  Count: 
  0
  <button
    id="increment"
    onClick={[Function]}
  >
    +
  </button>
  <button
    id="decrement"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;
